src/rxzsets.c

```diff
@@ -189,6 +191,129 @@ int ZAddCappedGenericCommand(RedisModuleCtx *ctx, RedisModuleString **argv,
+*/
+int ZUnionTopKCommand(RedisModuleCtx *ctx, RedisModuleString **argv, int argc) {
+  if (argc < 4) return RedisModule_WrongArity(ctx);
+
+  RedisModule_AutoMemory(ctx);
+  int rev = !strcasecmp("zunionmaxk", RedisModule_StringPtrLen(argv[0], NULL));
+
+  long long k;
+  if (RedisModule_StringToLongLong(argv[1], &k) != REDISMODULE_OK || k < 1) {
+    RedisModule_ReplyWithError(ctx, "ERR invalid k");
+    return REDISMODULE_ERR;
+  }
+
+  long long numkeys;
+  if (RedisModule_StringToLongLong(argv[2], &numkeys) != REDISMODULE_OK || numkeys < 1) {
+    RedisModule_ReplyWithError(ctx, "ERR invalid numkeys");
+    return REDISMODULE_ERR;
+  }
+
+  int has_weights = 0, with_scores = 0;
+  if (argc < 3 + numkeys) {
+    RedisModule_ReplyWithError(ctx, "ERR wrong key nums");
+    return REDISMODULE_ERR;
+  } else if (argc > 3 + numkeys) {
+    has_weights = !strcasecmp("weights", RedisModule_StringPtrLen(argv[3 + numkeys], NULL));
+    with_scores = !strcasecmp("withscores", RedisModule_StringPtrLen(argv[argc - 1], NULL));
+  }
+
+  // validate argc
+  if (has_weights) {
+    if (argc != 4 + 2 * numkeys + with_scores) {
+      return RedisModule_WrongArity(ctx);
+    }
+  } else {
+    if (argc != 3 + numkeys + with_scores) {
+      return RedisModule_WrongArity(ctx);
+    }
+  }
+
+  Vector *v = NewVector(ZsetEntry, numkeys);
+  for (size_t i = 0; i < numkeys; i++) {
+    RedisModuleKey *key = RedisModule_OpenKey(ctx, argv[3 + i], REDISMODULE_READ);
+    if (RedisModule_KeyType(key) == REDISMODULE_KEYTYPE_ZSET) {
+      if (RedisModule_ValueLength(key) == 0) {
+        continue;
+      }
+      (rev ? RedisModule_ZsetLastInScoreRange : RedisModule_ZsetFirstInScoreRange)(
+              key, REDISMODULE_NEGATIVE_INFINITE, REDISMODULE_POSITIVE_INFINITE, 0, 0);
+      ZsetEntry entry;
+      entry.key = key;
+      entry.element = RedisModule_ZsetRangeCurrentElement(key, &entry.score);
+      if (has_weights) {
+        if (RedisModule_StringToDouble(argv[4 + numkeys + i], &entry.weight) != REDISMODULE_OK) {
+          RedisModule_ReplyWithError(ctx, "ERR invalid weight");
+          Vector_Free(v);
+          return REDISMODULE_ERR;
+        }
+      } else {
+        entry.weight = 1;
+      }
+      __vector_PushPtr(v, &entry);
+    } else if (RedisModule_KeyType(key) == REDISMODULE_KEYTYPE_EMPTY) {
+      continue;
+    } else {
+      RedisModule_ReplyWithError(ctx, REDISMODULE_ERRORMSG_WRONGTYPE);
+      Vector_Free(v);
+      return REDISMODULE_ERR;
+    }
+  }
+  make_heap(v, 0, v->top, rev ? __zsetentry_less : __zsetentry_greater);
+
+  size_t reply_count = 0;
+  RedisModule_ReplyWithArray(ctx, REDISMODULE_POSTPONED_ARRAY_LEN);
+  for (size_t i = 0; i < k && v->top != 0; i++) {
+    // pop from heap
+    pop_heap(v, 0, v->top, rev ? __zsetentry_less : __zsetentry_greater);
+    ZsetEntry *entry = (ZsetEntry *) (v->data + ((v->top - 1) * v->elemSize));
```

Comment:
Here I just want to replace the `RedisModuleString *element` and the `double score` fields of the last item in the Vector. `Vector_Get(Vector *v, size_t pos, void *ptr);` will copy the value to ptr.

If I write `ZsetEntry entry; Vector_Get(v, pos, &entry);` here, the item in the Vector will be copied to the local variable (on the stack) and I should copy it back to the Vector after it is modified. This could be costly.